docs/concepts/parallelism.rst

```diff
@@ -0,0 +1,583 @@
+            return state.update(all_llm_outputs=all_llm_outputs)
+
+        def reads() -> List[str]:
+            return ["prompts"]
+
+        def writes() -> List[str]:
+            return ["all_llm_outputs"]
+
+Then, to run the application:
+
+.. code-block:: python
+
+    app = (
+        ApplicationBuilder()
+        .with_action(
+            prompt_generator=generate_prompts, # not defined above, this writes to prompts
+            multi_prompt_test=TestMultiplePrompts(),
+        ).with_transitions(
+            ("prompt_generator", "multi_prompt_test"),
+        )
+        .build()
+    )
+
+
+Run different actions over the same state
+-----------------------------------------
+
+
+For case (2) (mapping actions over the same state) you implement the ``MapActions`` class, doing the following:
+
+- We define a regular action ``query_llm_action()`` using the ``@action`` decorator. This takes in a model parameter (which we're going to bind later)
+- We also create a subclass of ``MapActions`` named ``TestMultipleModels``, which must implement ``.reads()``, ``.writes()``, ``.actions()``, ``.state()``, and ``.reduce()``.
+    - ``.reads()`` / ``.writes()`` define the state value it can interact with, just like the ``@action`` decorator
+    - ``.actions()`` leverages the ``query_llm_action()`` previously defined, binding with the different models we want to test
+    - ``.state()`` can read value from State and produces the state to pass to the actions produced by ``actions()``. In this case, it updates the prompt state value that's read by ``query_llm_action()``.
+    - ``.reduce()`` receives multiple states, one per result of the ``.actions()`` call, where the llm_output value is set by ``query_llm_action()`` in ``.actions()``. Then, it must set all_llm_output as specified in the ``MapStates.writes()`` method.
+- We pass an instance of the ``TestMultipleModels`` class to the ``ApplicationBuilder``, which will run the action over the states we provide.
+
+.. code-block:: python
+
+    from burr.core import action, state
+    from burr.core.parallelism import MapActions, RunnableGraph
+    from typing import Callable, Generator, List
+
+    @action(reads=["prompt", "model"], writes=["llm_output"])
+    def query_llm(state: State, model: str) -> State:
+        # TODO -- implement _query_my_llm to call litellm or something
+        return state.update(llm_output=_query_my_llm(prompt=state["prompt"], model=model))
+
+    class TestMultipleModels(MapActions):
+
+        def actions(self, state: State) -> Generator[Action | Callable | RunnableGraph, None, None]:
+            # make sure to add a name to the action
+            # This is not necessary for subgraphs, as actions will already have names
+            for action in [
+                query_llm.bind(model="gpt-4").with_name("gpt_4_answer"),
+                query_llm.bind(model="o1").with_name("o1_answer"),
+                query_llm.bind(model="claude").with_name("claude_answer"),
```

Comment:
The examples here use the same action with different parameters bound to it. Is it possible to instead map over different actions completely? For example something like this:

```python
@action(reads["query"], writes=["llm_output"])
def query_llm(state: State) -> State:
    return state.update(llm_output=some_function(state["query"]))

@action(reads["query"], writes=["llm_output"])
def count_tokens(state: State) -> State:
    return state.update(llm_output=another_function(state["query"]))

@action(reads["query"], writes=["llm_output"])
def draw_chart(state: State) -> State:
    return state.update(llm_output=a_third_function(state["query"]))

def actions(self, state: State) -> Generator[Action | Callable | RunnableGraph, None, None]:
    for action in [
        query_llm,
        count_tokens,
        draw_chart,
    ]:
        yield action
```

As a follow up question, is there a requirement that all the actions have to write the same keys? My naive intuition is that this is not a requirement, but I would have to do some extra work in the `reduce` function to make sure I get all available outputs:

```python
@action(reads["query"], writes=["llm_output"])
def query_llm(state: State) -> State:
    ..
@action(reads["query"], writes=["token_count"])
def count_tokens(state: State) -> State:
    ..
@action(reads["query"], writes=["chart_bytes"])
def draw_chart(state: State) -> State:
    ..

def reduce(self, states: Generator[State, None, None]) -> State:
    outputs = {}
    for state in states:
        if "llm_output" in state:
            outputs["llm_output"] = state["llm_output"]
        if "token_count" in state:
            outputs["token_count"] = state["token_count"]
        if "chart_bytes" in state:
            outputs["chart_bytes"] = state["chart_bytes"]
        return state.update(outputs)

def writes() -> List[str]:
    return ["llm_output", "token_count", "chart_bytes"]
```

Alternatively, perhaps it would be better to do this to avoid burr complaining that a particular state key was not present in the output
```python
def reduce(self, ..):
    outputs = {}
    ..
    return state.update({"outputs": outputs})

def writes():
    return ["outputs"]
```

Sorry for the long comment, just trying to wrap my head around the possibilities with the new API!

Reply:
Hey! Yes, I think that's supported (If I understand correctly). It's a little nuanced (and I could explain it better), but the idea is that the generator can yield an `Action` object, a `Callable` (function that will get turned into an action, or a `RunnableGraph` (a subgraph that functions as an action). Specifically, this allows you to do exactly what you want/mix + match.

The way to reconcile this with the example (with multiple `.bind()`) calls is that the `.bind` creates a new action, so for the sake of the `actions` generator it doesn't really matter whether it was created through `.bind(...)` or created through specifying another action entirely!

Regarding the follow-up, currently, yes, it's necessary to apply defaults. This is not a feature of parallelism, rather a feature of Burr. Specifically, we want some strict sense of what an action writes so we have guarentees on what should show up next. The general approach is to apply defaults -- E.G. say if state doesn't contain `foo`, it'll be `None` -- I think the simplest in your case would be to do something like:

```python
def reduce(self, states: Generator[State, None, None]) -> State:
    outputs = {'llm_output' : None, 'token_count' : None, 'chart_bytes' : None}  # initialize so you can overwrite with results
    # This is generic, or just have it like you did above, easier to read maybe
    for state in states:
        for key in list(outputs):
            if key in state:
                output[key] = state[key]
        return state.update(outputs)
```

Reply:
Amazing, thank you! This new API is going to be very helpful for us :)

Reply:
Great! Would love to hear about how you'd be using it/specifics (only if you feel comfortable sharing). This can help us get a sense of ergonomics (and we can send you test versions if you're interested in messing with it). Feel free to join the discord and DM us if that would be useful!